### ISSUE TITLE
[FEATURE/ISSUE] predefined structs for things like On "message" responses

### ISSUE THREAD

**Issue by blacktop** [frida/frida-go#7]
I'm testing out a script and trying to pretty parse the messages.  I'm not sure how consistant the fields are?  I'm assume not at all due to your choice of using `interface{}`

However, it seems pretty consistant at least for message "types" `send` and `error` etc etc So I thought it'd be awesome to get the response as a Message object?  I tried to do it myself except there is a weird issue I'm hitting:

```go
		type arg struct {
			TypeString        string `json:"typeString,omitempty"`
			TypeDescription   string `json:"typeDescription,omitempty"`
			Object            string `json:"object,omitempty"`
			ObjectDescription string `json:"objectDescription,omitempty"`
		}

		type Payload struct {
			TargetType         string `json:"targetType,omitempty"`
			TargetClass        string `json:"targetClass,omitempty"`
			TargetMethod       string `json:"targetMethod,omitempty"`
			Args               []arg  `json:"args,omitempty"`
			ReturnType         string `json:"returnType,omitempty"`
			RetTypeDescription string `json:"retTypeDescription,omitempty"`
			ReturnDescription  string `json:"returnDescription,omitempty"`
		}

		type Message struct {
			Type         string  `json:"type,omitempty"`
			Payload      Payload `json:"payload,omitempty"`
			Description  string  `json:"description,omitempty"`
			Stack        string  `json:"stack,omitempty"`
			FileName     string  `json:"fileName,omitempty"`
			LineNumber   int     `json:"lineNumber,omitempty"`
			ColumnNumber int     `json:"columnNumber,omitempty"`
		}

		type outterMsg struct {
			Type    string `json:"type,omitempty"`
			Payload string `json:"payload,omitempty"`
		}

		script.On("message", func(data string) {
			fmt.Println(data)
			var msg outterMsg
			var payload Payload
			if err := json.Unmarshal([]byte(data), &msg); err != nil {
				log.Warnf("Error parsing message: %v", err)
			}
			if err := json.Unmarshal([]byte(msg.Payload), &payload); err != nil {
				log.Warnf("Error parsing message: %v", err)
			}
			log.Infof("Received '%s' - %v", msg.Type, payload)
		})
```

So I have to break it into two parts because the data is:
```bash
"{\"type\":\"send\",\"payload\":\"{\\\"targetType\\\":\\\"-\\\",\\\"targetClass\\\":\\\"NSRegularExpression\\\",\\\"targetMethod\\\":\\\"copyWithZone:\\\",\\\"args\\\":[{\\\"typeString\\\":\\\"@\\\",\\\"typeDescription\\\":\\\"UIWindowLayer *\\\",\\\"object\\\":\\\"0x21234123\\\",\\\"objectDescription\\\":\\\"/* <UIWindowLayer: 0x21234123 > */ 0x21234123\\\"},{\\\"typeString\\\":\\\":\\\",\\\"typeDescription\\\":\\\"SEL\\\",\\\"object\\\":\\\"0x21234123\\\",\\\"objectDescription\\\":\\\"/* retainWeakReference */ 0x21234123\\\"}],\\\"returnType\\\":\\\"B\\\",\\\"retTypeDescription\\\":\\\"bool\\\",\\\"returnDescription\\\":\\\"0x21234123\\\"}\"}"
```		

It's like the type is one level of JSON marshaling and the payload is a 2nd level of marshaling notice the increase in `\`

I'd like to just use the `Message` struct to Unmarshal to, however, if I do I get the `type` but it fails on the `Payload` so I have to treat it as another []byte(string) and then it works.

SO IN SUMMARY:

- It might be nice to have the `Message` struct be included in **frida-go** which could include all the types "send", "error" etc
- There might be an issue in how your are building the script.On "message" data where the payload is getting EXTRA json marshalling?

Thanks again!

**Comment by NSEcho** [comment]
Hi again. Some types are always there for example `type` field and the payload field is of course dependent on the script. Message callback populate the `gchar*` which is equivalent to the `string` in go and populates the `gbytes*` as second param which gets converted to the `[]byte`.  

I can make some helper func which would receive the string and return the `Message` struct. Let me know if that sounds okay and it would be pretty much close to what you did already.

**Comment by NSEcho** [comment]
I made a new release `v0.4.1` with `Message` type and func `ScriptMessageToMessage` receiving `message` string from the callback.

Small example:


```golang
package main

import (
	"bufio"
	"fmt"
	"github.com/frida/frida-go/frida"
	"os"
)

var script = `
var invocation = {
"targetType": "someTargetType",
"targetClass": "someClass",
"targetMethod": "someMethod",
"args": []
}

send(JSON.stringify(invocation));
send("raw string data");
`

func main() {
	sess, err := frida.Attach(0)
	if err != nil {
		panic(err)
	}

	sc, err := sess.CreateScript(script)
	if err != nil {
		panic(err)
	}
	sc.On("message", func(message string) {
		m, err := frida.ScriptMessageToMessage(message)
		if err != nil {
			panic(err)
		}
		switch m.Type {
		case frida.MessageTypeLog:
			switch m.Level {
			case frida.LevelTypeLog:
				fmt.Println("Got log info")
			case frida.LevelTypeError:
				fmt.Println("Got log error")
			}
		case frida.MessageTypeSend:
			if m.IsPayloadMap {
				for k, v := range m.Payload.(map[string]interface{}) {
					fmt.Printf("%s => %s\n", k, v)
				}
			} else {
				fmt.Printf("Payload is string: %s\n", m.Payload)
			}
		}
	})
	sc.Load()

	// sc.ExportsCall("hook", "*[NSMutable* initWith*]")

	r := bufio.NewReader(os.Stdin)
	r.ReadLine()
}
```

Running this, gets us:

```bash
$ go build -o main.go && ./main
targetClass => someClass
targetMethod => someMethod
args => []
targetType => someTargetType
Payload is string: raw string data
```

**Comment by blacktop** [comment]
That's awesome!! Can't wait to try it out ❤️

**Comment by NSEcho** [comment]
Great, let me know if that is okay.

**Comment by blacktop** [comment]
@lateralusd I'm not sure what changed but now I can't supply a valid device id in the call to: 
```go
			dev, err = mgr.DeviceByID(udid)
			if err != nil {
				return fmt.Errorf("failed to get device by id %s: %v", udid, err)
			}
```
I keep getting the error: `failed to get device by id <MY ID>: FError: Device not found`		

This worked before?

**Comment by NSEcho** [comment]
Nothing has been changed, that is just the way frida works. The only devices you are going to have prior to enumerating devices is `local` and `socket`. First call to the `manager.EnumerateDevices()` will populate the list with other devices availlable. So in order for the example to work, call `manager.EnumerateDevices()` prior to getting a specific device.

**Comment by blacktop** [comment]
Ok so I see you didn't change anything that would affect this HOWEVER I did in that before I was run `EnumerateDevices` and THEN I could call `mgr.DeviceByID` and it would work.  It's as if `EnumerateDevices` allocates the info or something so `DeviceByID` won't work unless `EnumerateDevices` is called first?  This is probably NOT the behavior you want?

I also just tried `FindDeviceByID` instead of `DeviceByID` and it didn't error, however, when I tried to run `Name()` on the returned device it crashes with `"unexpected signal during runtime execution"` ??

**Comment by blacktop** [comment]
So I guess you prob want to actually run the C code for `EnumerateDevices` inside both `FindDeviceByID` and `DeviceByID`? Seems kinda gross, but maybe it's what the C code is expecting to happen? Or maybe `EnumerateDevices` should happen in the initializer for `NewDeviceManager` which prob makes the most sense?

**Comment by NSEcho** [comment]
I am going to add the `frida.DeviceByID(id)` with the default manager. For all the rest of the methods on the `Manager` type, `EnumerateDevices` should be called.

**Comment by blacktop** [comment]
Ya I think it's reasonable for the user to expect that once they have a manager they should be able to call ANY of the it's functions without errors or crashes and without having to call other setup functions?

Also if they call any of the Device getters they devices objects they get shouldn't be partially initialized so if you call a property getter on them it won't crash.

**Comment by blacktop** [comment]
@lateralusd wouldn't changing `NewDeviceManager ` to:
```go
// NewDeviceManager returns new frida device manager.
func NewDeviceManager() *DeviceManager {
	manager := C.frida_device_manager_new()
	mgr := &DeviceManager{manager}
        mgr.EnumerateDevices()
	return mgr
}
```
fix all these issues?

**Comment by blacktop** [comment]
Oh and my apologies for using this issue for these others.  For the initial purpose of the issue:  I've used the NEW `frida.ScriptMessageToMessage` and it is awesome! Thank you!!

**Comment by blacktop** [comment]
Not sure if this is my OCD, but should `	LevelTypeLog   LevelType = "info"` be `	LevelTypeInfo   LevelType = "info"` ?

**Comment by NSEcho** [comment]
> Ya I think it's reasonable for the user to expect that once they have a manager they should be able to call ANY of the it's functions without errors or crashes and without having to call other setup functions?

For the first part, I am trying to keep in sync with `frida-python` as much as I can. If someone wants to quickly get access to the `local`, `socket` or `device by id`(now) the user can quickly do it with the default manager. It would kinda solve it but hides the underlying logic and makes a difference between go and python bindings which I try not to do. For the second part, `GetDeviceByID` will throw an error if it can't find the device, while the `FindDeviceByID` does not throw an error and instead returns null. 

> Also if they call any of the Device getters they devices objects they get shouldn't be partially initialized so if you call a property getter on them it won't crash.

For this part you are right, will get it sorted out.

> Not sure if this is my OCD, but should LevelTypeLog LevelType = "info" be LevelTypeInfo LevelType = "info" ?

`info` level is emitted on calling `console.log`, `warn` on `console.warn` and `error` on `console.error`.

Thank you for point out to these issues, in [v0.5.0](https://github.com/frida/frida-go/releases/tag/v0.5.0) you can now use:

```golang
dev, err := frida.DeviceById(id)
```

**Comment by blacktop** [comment]
Just my 2 cents, I understand the instinct to match `frida-python` as it is the MOST popular binding, however, that assumes all the users of your lib will be coming from python, I think it's more likely the users will be coming from the Go community, so you might want to adhere more to idiomatic Go patterns so it feels natural.  I have made a lot of CGO bindings myself and I know it's gross and like mixing two apposing worlds and you've done a tremendous job.  Also the addition of `frida.ScriptMessageToMessage` already, in my opinion, moves the lib firmly into the Go "magic" feel, which is awesome, rather than making users double-json-marshal or do weird string parsing so thank you!
